day1/day1.zig

```diff
@@ -160,49 +101,46 @@ fn part2(buf: []const u8) !void {
     const stdout = std.io.getStdOut().writer();
     var it = mem.tokenize(buf, ", \n");
 
+    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
+    defer _ = gpa.deinit();
```

Comment:
Is this semantically equivalent to `defer gpa.deinit();`?

Reply:
Yep, but you'll get a compiler warning if you do so, since `GeneralPurposeAllocator`'s `deinit` returns a value — [`true` if leaks were detected](https://github.com/ziglang/zig/blob/6fb105fdd7798dc988de09a7b6709c5168355dfa/lib/std/heap/general_purpose_allocator.zig#L335-L340). We don't need to check the return value because we'll get big traces showing which allocations leaked by default in `Debug` and `ReleaseSafe` compile modes.

Reply:
(Sorry I'm taking so long to merge this LOL... I'm working on it!)

Can you use `defer` to change the return value from a function from non-error to error type? If not, if you are in `Release` compile mode, I don't see how you can use the return value for anything other than to print a message since `defer` runs at end of scope _just_<sup>1</sup> before returning.

1. For some value of _just_ I don't know yet :).

Reply:
Zig blocks can be used in a surprisingly huge number of places (like that `absCast` example from email where one's used to calculate the return type of a function), and here is no exception; you could just as well do this:

```zig
    defer {
        const leaks_found = gpa.deinit();
        std.debug.print("leaks found? {}\n", .{leaks_found});
    }
```

You can't use defer to change the return value, though; you can't return from a defer (and so can't `try` expressions either). Generally if someone wants you to be able to change the return value based on the result of an expression like this, it'll return an error union instead of just a bool so you can `try` it where appropriate, but I guess they figure a leak being found isn't an error condition.

(In `ReleaseFast`/`ReleaseSmall` safety checks are disabled ([by default](https://github.com/ziglang/zig/blob/6fb105fdd7798dc988de09a7b6709c5168355dfa/lib/std/heap/general_purpose_allocator.zig#L146)) anyway, so it'll always return false if that's kept that way.)